src/manager.ts

```diff
@@ -1,550 +1,426 @@
-				file =
-					this._workspaceManager.getFileRelativeToDefaultWorkspace(file) ?? '';
-			}
-		}
-		return file;
-	}
-
-	/**
-	 * @description notify the user that they are opening a file outside the current workspace(s).
-	 */
-	private notifyLooseFileOpen(): void {
-		/* __GDPR__
-			"preview.fileOutOfWorkspace" : {}
-		*/
-		this._reporter.sendTelemetryEvent('preview.fileOutOfWorkspace');
-		if (
-			!this._notifiedAboutLooseFiles &&
-			SettingUtil.GetConfig(this._extensionUri).notifyOnOpenLooseFile
-		) {
-			vscode.window
-				.showWarningMessage(
-					localize(
-						'notPartOfWorkspace',
-						'Previewing a file that is not a child of the server root. To see fully correct relative file links, please open a workspace at the project root.'
-					),
-					DONT_SHOW_AGAIN
-				)
-				.then((selection: vscode.MessageItem | undefined) => {
-					if (selection == DONT_SHOW_AGAIN) {
-						SettingUtil.UpdateSettings(Settings.notifyOnOpenLooseFile, false);
-					}
-				});
-		}
-		this._notifiedAboutLooseFiles = true;
-	}
-}
+import { Disposable } from './utils/dispose';
+import * as vscode from 'vscode';
+import { PathUtil } from './utils/pathUtil';
+import TelemetryReporter from 'vscode-extension-telemetry';
+import { ConnectionManager } from './connectionInfo/connectionManager';
+import { BrowserPreview } from './editorPreview/browserPreview';
+import { SETTINGS_SECTION_ID } from './utils/settingsUtil';
+import * as nls from 'vscode-nls';
+import { ServerTaskProvider } from './task/serverTaskProvider';
+import { EndpointManager } from './infoManagers/endpointManager';
+import { PreviewManager } from './editorPreview/previewManager';
+import { Connection } from './connectionInfo/connection';
+import { existsSync } from 'fs';
+import { StatusBarNotifier } from './server/serverUtils/statusBarNotifier';
+import { LIVE_PREVIEW_SERVER_ON } from './utils/constants';
+import { ServerManager } from './server/serverManager';
+
+const localize = nls.loadMessageBundle();
+
+/**
+ * This object re-serializes the webview after a reload
+ */
+class PanelSerializer
+	extends Disposable
+	implements vscode.WebviewPanelSerializer
+{
+	private readonly _onShouldRevive = this._register(
+		new vscode.EventEmitter<{ webviewPanel: vscode.WebviewPanel; state: any }>()
+	);
+
+	public readonly onShouldRevive = this._onShouldRevive.event;
+
+	deserializeWebviewPanel(
+		webviewPanel: vscode.WebviewPanel,
+		state: any
+	): Thenable<void> {
+		// fire event to parent, since all info needed to re-open a panel is in the parent
+		this._onShouldRevive.fire({ webviewPanel, state });
+		return Promise.resolve();
+	}
+}
+
+/**
+ * `Manager` is a singleton instance that managers all of the servers, the previews, connection info, etc.
+ * It also facilitates opening files (sometimes by calling `PreviewManager`) and starting the associated servers.
+ */
+export class Manager extends Disposable {
+	private _serverManagers: Map<vscode.Uri | undefined, ServerManager>;
```

Comment:
Same comment about URIs as map keys here

Reply:
It throws me off that all of the Managers are singletons except ServerManager

Reply:
changed `ServerManager` to `ServerGrouping`, which should make things less unsettling 😅😅